File: Week_01/java/81_search_in_rotated_sorted_array_ii.java

class Solution {
    public boolean search(int[] nums, int target) {

        int left=0, right = nums.length - 1;
        int mid = (left - right) / 2 + right;

        while(left < right) {
            mid = (left - right) / 2 + right;
            if (target == nums[mid] || target == nums[left] || target == nums[right]) return true;

            if(nums[mid] > nums[left]) {
                if (target > nums[left] && target < nums[mid] ) right = mid - 1;
                else left = mid + 1;
            } else if(nums[mid] < nums[left]) {
                if (target > nums[mid] && target < nums[right]) left = mid + 1;
                else right = mid - 1;
            } else {
                left++;
            }
        }

        return nums[mid] == target;

    }
}

File: Week_01/python/81_search_in_rotated_sorted_array_ii.py
from typing import List


class Solution:
    def search(self, nums: List[int], target: int) -> bool:
        left, right = 0, len(nums) - 1
        mid = (left - right) // 2 + right

        while left < right:
            mid = (left - right) // 2 + right
            print(left, " ", mid, " ", right, " ")

            if target in [nums[mid],  nums[left], nums[right]]: return True

            if nums[mid] > nums[left]:
                if nums[mid] > target > nums[left]: right = mid - 1
                else: left = mid + 1
            elif nums[mid] < nums[left]:
                if nums[mid] < target < nums[right]: left = mid + 1
                else: right = mid - 1
            else:
                # 点睛之笔，如果mid和left的同值，则left++
                left += 1

        return nums[mid] == target

